File: android-lib/src/androidTest/java/com/kinvey/androidTest/store/datastore/LiveServiceTest.java
package com.kinvey.androidTest.store.datastore;

import android.content.Context;
import android.util.Log;

import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.filters.SmallTest;
import androidx.test.platform.app.InstrumentationRegistry;

import com.kinvey.android.Client;
import com.kinvey.android.store.DataStore;
import com.kinvey.androidTest.TestManager;
import com.kinvey.androidTest.model.LiveModel;
import com.kinvey.java.store.BaseUserStore;
import com.kinvey.java.store.KinveyDataStoreLiveServiceCallback;
import com.kinvey.java.store.KinveyLiveServiceStatus;
import com.kinvey.java.store.LiveServiceRouter;
import com.kinvey.java.store.StoreType;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.io.IOException;
import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;

import static com.kinvey.androidTest.TestManager.PASSWORD;
import static com.kinvey.androidTest.TestManager.USERNAME;
import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertTrue;


/**
 * Created by yuliya on 2/19/17.
 */

@RunWith(AndroidJUnit4.class)
@SmallTest
public class LiveServiceTest {

    private static final int DEFAULT_TIMEOUT = 60 * 1000;
    private static final String TAG = "LiveServiceTest";

    private Client client;
    private TestManager<LiveModel> testManager;
    private DataStore<LiveModel> store;

    @Before
    public void setUp() throws InterruptedException, IOException {
        Context mMockContext = InstrumentationRegistry.getInstrumentation().getTargetContext();
        client = new Client.Builder(mMockContext).build();
        client.enableDebugLogging();
        testManager = new TestManager<>();
        testManager.login(USERNAME, PASSWORD, client);
    }

    @After
    public void tearDown() {
        client.performLockDown();
        if (client.getKinveyHandlerThread() != null) {
            try {
                client.stopKinveyHandlerThread();
            } catch (Throwable throwable) {
                throwable.printStackTrace();
            }
        }
    }


    @Test
    public void testRegisterUnregisterSync() throws InterruptedException, IOException {
        assertTrue(client.isUserLoggedIn());
        BaseUserStore.registerLiveService();
        assertTrue(LiveServiceRouter.getInstance().isInitialized());
        BaseUserStore.unRegisterLiveService();
        assertFalse(LiveServiceRouter.getInstance().isInitialized());
    }

    @Test
    public void testSubscribeUnsubscribeSync() throws InterruptedException, IOException {
        assertTrue(client.isUserLoggedIn());
        BaseUserStore.registerLiveService();
        assertTrue(LiveServiceRouter.getInstance().isInitialized());
        store = DataStore.collection(LiveModel.COLLECTION, LiveModel.class, StoreType.SYNC, client);
        LiveModel model = new LiveModel();
        model.setUsername("Live model name");
        store.save(model);
        store.pushBlocking();
        store.subscribe(new KinveyDataStoreLiveServiceCallback<LiveModel>() {
            @Override
            public void onNext(LiveModel next) {
                Log.d(TAG,  next.toString());
            }

            @Override
            public void onError(Exception e) {
                Log.d(TAG,  e.toString());
            }

            @Override
            public void onStatus(KinveyLiveServiceStatus status) {
                Log.d(TAG,  status.toString());
            }
        });
        store.unsubscribe();
        BaseUserStore.unRegisterLiveService();
        assertFalse(LiveServiceRouter.getInstance().isInitialized());
    }

/*
    @Test
    public void testSubscribe() throws InterruptedException, IOException {
        assertTrue(client.isUserLoggedIn());
        BaseUserStore.registerLiveService();
        assertTrue(LiveServiceRouter.getInstance().isInitialized());
        store = DataStore.collection(LiveModel.COLLECTION, LiveModel.class, StoreType.SYNC, client);
        DataStore<LiveModel> networkStore = DataStore.collection(LiveModel.COLLECTION, LiveModel.class, StoreType.NETWORK, client);
        LiveModel model = new LiveModel();
        model.setUsername("Live model name");
        LiveModel savedModel = store.save(model);
        store.pushBlocking();
        testManager = new TestManager<>();


        final CountDownLatch latch = new CountDownLatch(1);
        CustomKinveyLiveServiceCallback<LiveModel> liveServiceCallback = new CustomKinveyLiveServiceCallback<>(latch);
        CustomKinveyClientCallback<Boolean> subscribeCallback = testManager.subscribe(store, liveServiceCallback);
        assertTrue(subscribeCallback.getResult());
        assertNull(subscribeCallback.getError());
        savedModel.setUsername("Live model name - changed");
        testManager.saveCustomInBackground(networkStore, savedModel);
        System.out.println("stopped");
        latch.await(20, TimeUnit.SECONDS);

        assertNotNull(liveServiceCallback.getResult());
        assertNull(liveServiceCallback.getError());
        assertEquals("Live model name - changed", liveServiceCallback.getResult().getUsername());
        store.unsubscribe();
        BaseUserStore.unRegisterLiveService();
        assertFalse(LiveServiceRouter.getInstance().isInitialized());
        networkStore.delete(liveServiceCallback.getResult().getId());
    }
*/

    @Test
    public void testDeviceUuid() throws ClassNotFoundException, NoSuchMethodException, IllegalAccessException, InvocationTargetException, InstantiationException {
        Class<?> aClass = Class.forName("com.kinvey.android.UuidFactory");
        Constructor<?> constructor = aClass.getDeclaredConstructor(Context.class);
        constructor.setAccessible(true);
        Object uuidFactory = constructor.newInstance(client.getContext());
        String theFirstUiId = uuidFactory.getClass().getDeclaredMethod("getDeviceUuid").invoke(uuidFactory).toString();
        Object uuidFactorySecond = constructor.newInstance(client.getContext());
        String theSecondUiId = uuidFactorySecond.getClass().getDeclaredMethod("getDeviceUuid").invoke(uuidFactorySecond).toString();
        assertEquals(theFirstUiId, theSecondUiId);
    }

    @Test
    public void testLogout() throws InterruptedException, IOException {
        assertTrue(client.isUserLoggedIn());
        BaseUserStore.registerLiveService();
        assertTrue(LiveServiceRouter.getInstance().isInitialized());
        store = DataStore.collection(LiveModel.COLLECTION, LiveModel.class, StoreType.SYNC, client);
        store.subscribe(new KinveyDataStoreLiveServiceCallback<LiveModel>() {
            @Override
            public void onNext(LiveModel next) {
                Log.d(TAG,  next.toString());
            }

            @Override
            public void onError(Exception e) {
                Log.d(TAG,  e.toString());
            }

            @Override
            public void onStatus(KinveyLiveServiceStatus status) {
                Log.d(TAG,  status.toString());
            }
        });

        testManager = new TestManager<>();
        BaseUserStore.logout(client);
        assertFalse(LiveServiceRouter.getInstance().isInitialized());
    }

}

File: android-lib/src/androidTest/java/com/kinvey/androidTest/store/datastore/LiveServiceTest.kt
package com.kinvey.androidTest.store.datastore

import android.content.Context
import android.util.Log
import androidx.test.filters.SmallTest
import androidx.test.platform.app.InstrumentationRegistry
import androidx.test.runner.AndroidJUnit4
import com.kinvey.android.Client
import com.kinvey.android.Client.Builder
import com.kinvey.android.model.User
import com.kinvey.android.store.DataStore
import com.kinvey.android.store.DataStore.Companion.collection
import com.kinvey.androidTest.TestManager
import com.kinvey.androidTest.TestManager.Companion.PASSWORD
import com.kinvey.androidTest.TestManager.Companion.TEST_USERNAME
import com.kinvey.androidTest.TestManager.Companion.USERNAME
import com.kinvey.androidTest.model.LiveModel
import com.kinvey.java.AbstractClient
import com.kinvey.java.dto.BaseUser
import com.kinvey.java.store.BaseUserStore.logout
import com.kinvey.java.store.BaseUserStore.registerLiveService
import com.kinvey.java.store.BaseUserStore.unRegisterLiveService
import com.kinvey.java.store.KinveyDataStoreLiveServiceCallback
import com.kinvey.java.store.KinveyLiveServiceStatus
import com.kinvey.java.store.LiveServiceRouter.Companion.instance
import com.kinvey.java.store.StoreType
import junit.framework.Assert.*
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import java.io.IOException
import java.lang.reflect.Constructor
import java.lang.reflect.InvocationTargetException

/**
 * Created by yuliya on 2/19/17.
 */

@RunWith(AndroidJUnit4::class)
@SmallTest
class LiveServiceTest {

    private var client: Client<*>? = null
    private var testManager: TestManager<LiveModel>? = null
    private var store: DataStore<LiveModel>? = null

    @Before
    @Throws(InterruptedException::class, IOException::class)
    fun setUp() {
        val mMockContext: Context? = InstrumentationRegistry.getInstrumentation().targetContext
        client = Builder<User>(mMockContext).build()
        client?.enableDebugLogging()
        testManager = TestManager()
        testManager?.login(USERNAME, PASSWORD, client)
    }

    @After
    fun tearDown() {
        client?.performLockDown()
        if (Client.kinveyHandlerThread != null) {
            try {
                client?.stopKinveyHandlerThread()
            } catch (throwable: Throwable) {
                throwable.printStackTrace()
            }
        }
    }

    @Test
    @Throws(InterruptedException::class, IOException::class)
    fun testRegisterUnregisterSync() {
        assertTrue(client?.isUserLoggedIn == true)
        registerLiveService<BaseUser>()
        assertTrue(instance?.isInitialized == true)
        unRegisterLiveService<BaseUser>()
        assertFalse(instance?.isInitialized == true)
    }

    @Test
    @Throws(InterruptedException::class, IOException::class)
    fun testSubscribeUnsubscribeSync() {
        assertTrue(client?.isUserLoggedIn == true)
        registerLiveService<BaseUser>()
        assertTrue(instance?.isInitialized == true)
        store = collection(LiveModel.COLLECTION, LiveModel::class.java, StoreType.SYNC, client)
        val model = LiveModel()
        model.username = "Live model name"
        store?.save(model)
        store?.pushBlocking()
        store?.subscribe(object : KinveyDataStoreLiveServiceCallback<LiveModel> {
            override fun onNext(next: LiveModel) {
                Log.d(TAG, next.toString())
            }
            override fun onError(e: Exception) {
                Log.d(TAG, e.toString())
            }
            override fun onStatus(status: KinveyLiveServiceStatus) {
                Log.d(TAG, status.toString())
            }
        })
        store?.unsubscribe()
        unRegisterLiveService<BaseUser>()
        assertFalse(instance?.isInitialized == true)
    }

    @Test
    @Throws(ClassNotFoundException::class, NoSuchMethodException::class, IllegalAccessException::class,
            InvocationTargetException::class, InstantiationException::class)
    fun testDeviceUuid() {
        val aClass = Class.forName("com.kinvey.android.UuidFactory")
        val constructor = aClass.getDeclaredConstructor(Context::class.java)
        constructor.isAccessible = true
        val uuidFactory: Any = constructor.newInstance(client?.context)
        val theFirstUiId = uuidFactory.javaClass.getDeclaredMethod("getDeviceUuid").invoke(uuidFactory).toString()
        val uuidFactorySecond: Any = constructor.newInstance(client?.context)
        val theSecondUiId = uuidFactorySecond.javaClass.getDeclaredMethod("getDeviceUuid").invoke(uuidFactorySecond).toString()
        assertEquals(theFirstUiId, theSecondUiId)
    }

    @Test
    @Throws(InterruptedException::class, IOException::class)
    fun testLogout() {
        assertTrue(client?.isUserLoggedIn == true)
        registerLiveService<BaseUser>()
        assertTrue(instance?.isInitialized == true)
        store = collection(LiveModel.COLLECTION, LiveModel::class.java, StoreType.SYNC, client)
        store?.subscribe(object : KinveyDataStoreLiveServiceCallback<LiveModel> {
            override fun onNext(next: LiveModel) {
                Log.d(TAG, next.toString())
            }
            override fun onError(e: Exception) {
                Log.d(TAG, e.toString())
            }
            override fun onStatus(status: KinveyLiveServiceStatus) {
                Log.d(TAG, status.toString())
            }
        })
        testManager = TestManager()
        logout<BaseUser>(client as AbstractClient<BaseUser>)
        assertFalse(instance?.isInitialized == true)
    }

    companion object {
        private const val DEFAULT_TIMEOUT = 60 * 1000
        private const val TAG = "LiveServiceTest"
    }
}